File: packages/dotnet/OpenSkp/StlExport.cs
using System;
using System.Globalization;
using System.IO;
using System.Text;

namespace OpenSkp
{
    /// <summary>
    /// STL (Standard Triangle Language) exporter for baked <see cref="Scene"/> objects.
    /// Supports both ASCII and Binary STL formats.
    /// </summary>
    public static class StlExport
    {
        private static (float nx, float ny, float nz) CalculateNormal(
            (float x, float y, float z) v0,
            (float x, float y, float z) v1,
            (float x, float y, float z) v2)
        {
            float e1x = v1.x - v0.x;
            float e1y = v1.y - v0.y;
            float e1z = v1.z - v0.z;

            float e2x = v2.x - v0.x;
            float e2y = v2.y - v0.y;
            float e2z = v2.z - v0.z;

            float nx = e1y * e2z - e1z * e2y;
            float ny = e1z * e2x - e1x * e2z;
            float nz = e1x * e2y - e1y * e2x;

            float len = (float)Math.Sqrt(nx * nx + ny * ny + nz * nz);
            if (len > 1e-12f)
            {
                return (nx / len, ny / len, nz / len);
            }
            return (0.0f, 0.0f, 0.0f);
        }

        /// <summary>
        /// Serialize a baked <see cref="Scene"/> into ASCII STL text format.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <param name="scale">Scale factor (e.g. 1000.0f to convert metres to millimetres for 3D slicers).</param>
        /// <returns>The formatted ASCII STL text string.</returns>
        public static string ToStlAscii(Scene scene, float scale = 1.0f)
        {
            if (scene == null) throw new ArgumentNullException(nameof(scene));

            var sb = new StringBuilder();
            sb.AppendLine("solid OpenSKP_Model");

            foreach (var prim in scene.GlbPrimitives)
            {
                int triCount = prim.Indices.Length / 3;
                for (int i = 0; i < triCount; i++)
                {
                    uint i0 = prim.Indices[i * 3];
                    uint i1 = prim.Indices[i * 3 + 1];
                    uint i2 = prim.Indices[i * 3 + 2];

                    var v0 = (
                        x: prim.Positions[i0 * 3] * scale,
                        y: prim.Positions[i0 * 3 + 1] * scale,
                        z: prim.Positions[i0 * 3 + 2] * scale
                    );
                    var v1 = (
                        x: prim.Positions[i1 * 3] * scale,
                        y: prim.Positions[i1 * 3 + 1] * scale,
                        z: prim.Positions[i1 * 3 + 2] * scale
                    );
                    var v2 = (
                        x: prim.Positions[i2 * 3] * scale,
                        y: prim.Positions[i2 * 3 + 1] * scale,
                        z: prim.Positions[i2 * 3 + 2] * scale
                    );

                    var n = CalculateNormal(v0, v1, v2);

                    string nx = n.nx.ToString("F6", CultureInfo.InvariantCulture);
                    string ny = n.ny.ToString("F6", CultureInfo.InvariantCulture);
                    string nz = n.nz.ToString("F6", CultureInfo.InvariantCulture);

                    string v0x = v0.x.ToString("F6", CultureInfo.InvariantCulture);
                    string v0y = v0.y.ToString("F6", CultureInfo.InvariantCulture);
                    string v0z = v0.z.ToString("F6", CultureInfo.InvariantCulture);

                    string v1x = v1.x.ToString("F6", CultureInfo.InvariantCulture);
                    string v1y = v1.y.ToString("F6", CultureInfo.InvariantCulture);
                    string v1z = v1.z.ToString("F6", CultureInfo.InvariantCulture);

                    string v2x = v2.x.ToString("F6", CultureInfo.InvariantCulture);
                    string v2y = v2.y.ToString("F6", CultureInfo.InvariantCulture);
                    string v2z = v2.z.ToString("F6", CultureInfo.InvariantCulture);

                    sb.AppendLine($"  facet normal {nx} {ny} {nz}");
                    sb.AppendLine("    outer loop");
                    sb.AppendLine($"      vertex {v0x} {v0y} {v0z}");
                    sb.AppendLine($"      vertex {v1x} {v1y} {v1z}");
                    sb.AppendLine($"      vertex {v2x} {v2y} {v2z}");
                    sb.AppendLine("    endloop");
                    sb.AppendLine("  endfacet");
                }
            }

            sb.AppendLine("endsolid OpenSKP_Model");
            return sb.ToString();
        }

        /// <summary>
        /// Serialize a baked <see cref="Scene"/> into Little-Endian Binary STL format.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <param name="scale">Scale factor (e.g. 1000.0f for mm).</param>
        /// <returns>Packed Little-Endian Binary STL byte array.</returns>
        public static byte[] ToStlBinary(Scene scene, float scale = 1.0f)
        {
            if (scene == null) throw new ArgumentNullException(nameof(scene));

            int totalTriangles = 0;
            foreach (var prim in scene.GlbPrimitives)
            {
                totalTriangles += prim.Indices.Length / 3;
            }

            int bufferSize = 80 + 4 + totalTriangles * 50;
            byte[] bytes = new byte[bufferSize];

            // Header (80 bytes)
            byte[] header = Encoding.ASCII.GetBytes("# OpenSKP Binary STL Export");
            Array.Copy(header, bytes, Math.Min(header.Length, 80));

            // Triangle Count (uint32 Little-Endian)
            byte[] countBytes = BitConverter.GetBytes((uint)totalTriangles);
            if (!BitConverter.IsLittleEndian) Array.Reverse(countBytes);
            Array.Copy(countBytes, 0, bytes, 80, 4);

            int offset = 84;
            foreach (var prim in scene.GlbPrimitives)
            {
                int triCount = prim.Indices.Length / 3;
                for (int i = 0; i < triCount; i++)
                {
                    uint i0 = prim.Indices[i * 3];
                    uint i1 = prim.Indices[i * 3 + 1];
                    uint i2 = prim.Indices[i * 3 + 2];

                    var v0 = (
                        x: prim.Positions[i0 * 3] * scale,
                        y: prim.Positions[i0 * 3 + 1] * scale,
                        z: prim.Positions[i0 * 3 + 2] * scale
                    );
                    var v1 = (
                        x: prim.Positions[i1 * 3] * scale,
                        y: prim.Positions[i1 * 3 + 1] * scale,
                        z: prim.Positions[i1 * 3 + 2] * scale
                    );
                    var v2 = (
                        x: prim.Positions[i2 * 3] * scale,
                        y: prim.Positions[i2 * 3 + 1] * scale,
                        z: prim.Positions[i2 * 3 + 2] * scale
                    );

                    var n = CalculateNormal(v0, v1, v2);

                    // Normal (3x Float32)
                    WriteFloatLE(bytes, offset, n.nx);
                    WriteFloatLE(bytes, offset + 4, n.ny);
                    WriteFloatLE(bytes, offset + 8, n.nz);

                    // Vertices (9x Float32)
                    WriteFloatLE(bytes, offset + 12, v0.x);
                    WriteFloatLE(bytes, offset + 16, v0.y);
                    WriteFloatLE(bytes, offset + 20, v0.z);

                    WriteFloatLE(bytes, offset + 24, v1.x);
                    WriteFloatLE(bytes, offset + 28, v1.y);
                    WriteFloatLE(bytes, offset + 32, v1.z);

                    WriteFloatLE(bytes, offset + 36, v2.x);
                    WriteFloatLE(bytes, offset + 40, v2.y);
                    WriteFloatLE(bytes, offset + 44, v2.z);

                    // Attribute byte count (Uint16)
                    bytes[offset + 48] = 0;
                    bytes[offset + 49] = 0;

                    offset += 50;
                }
            }

            return bytes;
        }

        private static void WriteFloatLE(byte[] buffer, int offset, float value)
        {
            byte[] fBytes = BitConverter.GetBytes(value);
            if (!BitConverter.IsLittleEndian) Array.Reverse(fBytes);
            Array.Copy(fBytes, 0, buffer, offset, 4);
        }

        /// <summary>
        /// Export a baked <see cref="Scene"/> directly to an STL file.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <param name="outputPath">Destination file path (.stl).</param>
        /// <param name="binary">If true, writes binary STL. Otherwise writes ASCII STL.</param>
        /// <param name="scale">Scale factor (e.g. 1000.0f for mm).</param>
        public static void ExportStl(Scene scene, string outputPath, bool binary = false, float scale = 1.0f)
        {
            if (string.IsNullOrEmpty(outputPath)) throw new ArgumentException("Output path cannot be null or empty", nameof(outputPath));

            string? dir = Path.GetDirectoryName(outputPath);
            if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
            {
                Directory.CreateDirectory(dir);
            }

            if (binary)
            {
                byte[] data = ToStlBinary(scene, scale);
                File.WriteAllBytes(outputPath, data);
            }
            else
            {
                string text = ToStlAscii(scene, scale);
                File.WriteAllText(outputPath, text, Encoding.UTF8);
            }
        }
    }
}

File: packages/dart/lib/src/stl_export.dart
import 'dart:io';
import 'dart:math' as math;
import 'dart:typed_data';

import 'scene.dart';

(double, double, double) _calculateNormal(
  (double, double, double) v0,
  (double, double, double) v1,
  (double, double, double) v2,
) {
  final e1x = v1.$1 - v0.$1;
  final e1y = v1.$2 - v0.$2;
  final e1z = v1.$3 - v0.$3;

  final e2x = v2.$1 - v0.$1;
  final e2y = v2.$2 - v0.$2;
  final e2z = v2.$3 - v0.$3;

  final nx = e1y * e2z - e1z * e2y;
  final ny = e1z * e2x - e1x * e2z;
  final nz = e1x * e2y - e1y * e2x;

  final len = math.sqrt(nx * nx + ny * ny + nz * nz);
  if (len > 1e-12) {
    return (nx / len, ny / len, nz / len);
  }
  return (0.0, 0.0, 0.0);
}

/// Convert a baked [Scene] into ASCII STL text format.
String toStlAscii(Scene scene, {double scale = 1.0}) {
  final buffer = StringBuffer();
  buffer.writeln('solid OpenSKP_Model');

  for (final prim in scene.glbPrimitives) {
    final triCount = prim.indices.length ~/ 3;
    for (var i = 0; i < triCount; i++) {
      final i0 = prim.indices[i * 3];
      final i1 = prim.indices[i * 3 + 1];
      final i2 = prim.indices[i * 3 + 2];

      final v0 = (
        prim.positions[i0 * 3] * scale,
        prim.positions[i0 * 3 + 1] * scale,
        prim.positions[i0 * 3 + 2] * scale,
      );
      final v1 = (
        prim.positions[i1 * 3] * scale,
        prim.positions[i1 * 3 + 1] * scale,
        prim.positions[i1 * 3 + 2] * scale,
      );
      final v2 = (
        prim.positions[i2 * 3] * scale,
        prim.positions[i2 * 3 + 1] * scale,
        prim.positions[i2 * 3 + 2] * scale,
      );

      final normal = _calculateNormal(v0, v1, v2);

      buffer.writeln('  facet normal ${normal.$1.toStringAsFixed(6)} ${normal.$2.toStringAsFixed(6)} ${normal.$3.toStringAsFixed(6)}');
      buffer.writeln('    outer loop');
      buffer.writeln('      vertex ${v0.$1.toStringAsFixed(6)} ${v0.$2.toStringAsFixed(6)} ${v0.$3.toStringAsFixed(6)}');
      buffer.writeln('      vertex ${v1.$1.toStringAsFixed(6)} ${v1.$2.toStringAsFixed(6)} ${v1.$3.toStringAsFixed(6)}');
      buffer.writeln('      vertex ${v2.$1.toStringAsFixed(6)} ${v2.$2.toStringAsFixed(6)} ${v2.$3.toStringAsFixed(6)}');
      buffer.writeln('    endloop');
      buffer.writeln('  endfacet');
    }
  }

  buffer.writeln('endsolid OpenSKP_Model');
  return buffer.toString();
}

/// Convert a baked [Scene] into Little-Endian Binary STL byte array.
Uint8List toStlBinary(Scene scene, {double scale = 1.0}) {
  var totalTriangles = 0;
  for (final prim in scene.glbPrimitives) {
    totalTriangles += prim.indices.length ~/ 3;
  }

  final bufferSize = 80 + 4 + totalTriangles * 50;
  final bytes = Uint8List(bufferSize);
  final bdata = ByteData.view(bytes.buffer);

  // Write 80-byte header
  const headerText = '# OpenSKP Binary STL Export';
  for (var i = 0; i < 80; i++) {
    bytes[i] = i < headerText.length ? headerText.codeUnitAt(i) : 0;
  }

  // Write triangle count (uint32 Little-Endian)
  bdata.setUint32(80, totalTriangles, Endian.little);

  var offset = 84;
  for (final prim in scene.glbPrimitives) {
    final triCount = prim.indices.length ~/ 3;
    for (var i = 0; i < triCount; i++) {
      final i0 = prim.indices[i * 3];
      final i1 = prim.indices[i * 3 + 1];
      final i2 = prim.indices[i * 3 + 2];

      final v0 = (
        prim.positions[i0 * 3] * scale,
        prim.positions[i0 * 3 + 1] * scale,
        prim.positions[i0 * 3 + 2] * scale,
      );
      final v1 = (
        prim.positions[i1 * 3] * scale,
        prim.positions[i1 * 3 + 1] * scale,
        prim.positions[i1 * 3 + 2] * scale,
      );
      final v2 = (
        prim.positions[i2 * 3] * scale,
        prim.positions[i2 * 3 + 1] * scale,
        prim.positions[i2 * 3 + 2] * scale,
      );

      final normal = _calculateNormal(v0, v1, v2);

      // Normal (3x Float32)
      bdata.setFloat32(offset, normal.$1, Endian.little);
      bdata.setFloat32(offset + 4, normal.$2, Endian.little);
      bdata.setFloat32(offset + 8, normal.$3, Endian.little);

      // Vertex 0 (3x Float32)
      bdata.setFloat32(offset + 12, v0.$1, Endian.little);
      bdata.setFloat32(offset + 16, v0.$2, Endian.little);
      bdata.setFloat32(offset + 20, v0.$3, Endian.little);

      // Vertex 1 (3x Float32)
      bdata.setFloat32(offset + 24, v1.$1, Endian.little);
      bdata.setFloat32(offset + 28, v1.$2, Endian.little);
      bdata.setFloat32(offset + 32, v1.$3, Endian.little);

      // Vertex 2 (3x Float32)
      bdata.setFloat32(offset + 36, v2.$1, Endian.little);
      bdata.setFloat32(offset + 40, v2.$2, Endian.little);
      bdata.setFloat32(offset + 44, v2.$3, Endian.little);

      // Attribute byte count (Uint16)
      bdata.setUint16(offset + 48, 0, Endian.little);

      offset += 50;
    }
  }

  return bytes;
}

/// Export a baked [Scene] to an STL file at [path].
void exportStl(Scene scene, String path, {bool binary = false, double scale = 1.0}) {
  final file = File(path);
  final dir = file.parent;
  if (!dir.existsSync()) {
    dir.createSync(recursive: true);
  }

  if (binary) {
    final data = toStlBinary(scene, scale: scale);
    file.writeAsBytesSync(data);
  } else {
    final text = toStlAscii(scene, scale: scale);
    file.writeAsStringSync(text);
  }
}
